File: aoc2/aoc2.py
# Advent of Code 2
lines = [list(map(int, line.split(" "))) for line in open("input.txt", "r")]

def isDirectional (floors):
    return all([a < b for (a,b) in zip(floors, floors[1:])]) or \
    all([a > b for (a,b) in zip(floors, floors[1:])])

def isGradual (floors):
    return all(1 <= abs(a-b) <= 3 for (a,b) in zip(floors, floors[1:]))

def safetyDampener (floors):
    return any([isDirectional(seq) and isGradual(seq) for seq in removeOne(floors)])

def removeOne(lst):
    return [lst[:i] + lst[i+1:] for i in range(len(lst))]

result1 = sum([isDirectional(seq) and isGradual(seq) for seq in lines])

result2 = sum([safetyDampener(seq) for seq in lines])

print(result1, result2)

File: aoc2/aoc2.hs
import System.IO

isGradual :: [Int] -> Bool
isGradual xs = all (\(a, b) -> abs (a-b) >= 1 && abs(a-b) <= 3) $ zip xs (tail xs)

isDirectional :: [Int] -> Bool
isDirectional xs = all (\(a,b) -> a>b || a<b) $ zip xs (tail xs)

main :: IO ()
main = do
    contents <- readFile "input.txt"
    let sets :: [[Int]] = map (map read. words) $ lines contents
    let result :: Int = length $ filter (\x -> isDirectional x && isGradual x) sets
    print result

